src/lightning/data/dataset.py

```diff
@@ -0,0 +1,89 @@
+import os
+import tempfile
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Optional, Tuple
+
+from torch.utils.data import Dataset as TorchDataset
+
+from lightning.data.authenticators import _DatasetBackend, LocalDatasetBackend, S3DatasetBackend
+from lightning.data.dataset_index import get_index
+from lightning.data.fileio import OpenCloudFileObj
+
+
+class LightningDataset(TorchDataset, ABC):
+    """Dataset wrapper for optimized dataloading.
+
+    Arguments:
+
+        data_source: path of data directory.
+
+        backend: current options are "s3" or "local"
+
+        path_to_index_file: path to index file that lists all file contents of the data_source.
+    """
+
+    def __init__(self, data_source: str, backend: str = "local", path_to_index_file: Optional[str] = None):
+        super().__init__()
+        self.data_source = data_source
+
+        if not path_to_index_file:
+            tmpdir = tempfile.mkdtemp()
+            path_to_index_file = os.path.join(tmpdir, "index.txt")
+
+        self.index_file = os.path.abspath(os.path.expandvars(os.path.expanduser(path_to_index_file)))
+
+        self.files = self.get_index()
+
+        self.authenticator = self._chose_authenticator(backend=backend)
```

Comment:
This is a bit weird because it says it's an authenticator, but we are returning a `Backend` object.

I'd call this one `_init_backend` (which we will skip in the future if `backend` is an object and not a string)